src/cryptonote_core/service_node_list.cpp

```diff
@@ -1298,28 +1381,40 @@ namespace service_nodes
       return;
     }
 
-    std::vector<crypto::public_key> const snode_list = get_service_nodes_pubkeys();
-    quorum_manager &manager                          = m_transient_state.quorum_states[height];
-    for (int type_int = 0; type_int < (int)quorum_type::count; type_int++)
+    // The two quorums here have different selection criteria: the entire checkpoint quorum and the
+    // state change *validators* want only active service nodes, but the state change *workers*
+    // (i.e. the nodes to be tested) also include decommissioned service nodes.  (Prior to v12 there
+    // are no decommissioned nodes, so this distinction is irrelevant for network concensus).
+    auto active_snode_list = m_transient_state.active_service_nodes_infos();
+    decltype(active_snode_list) decomm_snode_list;
+    if (hf_version >= cryptonote::network_version_12_checkpointing)
+      decomm_snode_list = m_transient_state.decommissioned_service_nodes_infos();
+
+    quorum_manager &manager = m_transient_state.quorum_states[height];
+    for (int type_int = 0; type_int < static_cast<int>(quorum_type::count); type_int++)
     {
       auto type             = static_cast<quorum_type>(type_int);
       size_t num_validators = 0, num_workers = 0;
-      auto quorum                                = std::make_shared<testing_quorum>();
-      std::vector<size_t> const pub_keys_indexes = generate_shuffled_service_node_index_list(snode_list, block_hash, type);
+      auto quorum           = std::make_shared<testing_quorum>();
+      std::vector<size_t> pub_keys_indexes;
 
-      if (type == quorum_type::deregister)
+      if (type == quorum_type::state_change)
       {
         if (hf_version >= cryptonote::network_version_9_service_nodes)
         {
-          num_validators             = std::min(pub_keys_indexes.size(), DEREGISTER_QUORUM_SIZE);
-          size_t num_remaining_nodes = pub_keys_indexes.size() - num_validators;
-          num_workers                = std::max(num_remaining_nodes/DEREGISTER_NTH_OF_THE_NETWORK_TO_TEST, std::min(DEREGISTER_MIN_NODES_TO_TEST, num_remaining_nodes));
+          size_t total_nodes         = active_snode_list.size() + decomm_snode_list.size();
+          num_validators             = std::min(active_snode_list.size(), STATE_CHANGE_QUORUM_SIZE);
+          pub_keys_indexes           = generate_shuffled_service_node_index_list(total_nodes, block_hash, type, num_validators, active_snode_list.size());
```

Comment:
I don't get this bit, presuming 10 validators in a list of 100 nodes.

Shuffle from index (sublist_size == num_validators) 11 -> 100. Shuffle the whole list (sublist_up_to == active_snode_list.size()).

Or even, why the need for a double shuffle? Does this produce the same results for backwards compatibility pre HF12?

Reply:
The double shuffle only applies when there are inactive (i.e. decommissioned) snodes in the list.  If there aren't, there is just a single shuffle, as currently occurs.

When there are, the situation is like this:

We need 10 validators from a list of 100 active nodes, but there are actually 105 nodes: the 100 active ones *plus* 5 decommissioned nodes.  We need the decommissioned nodes to be selectable into the set of workers (so that they get retested to see if they are still down), but *not* in the set of validators (since they have been removed from active service until recommissioned by peers).  Thus we need to first shuffle the first 100 to pick 10 out of the random 100 *active* nodes.

Then we need to shuffle those remaining 5 into the set of workers, so we reshuffle everything after the validators, now *including* the extra 5 positions.  (Shuffling them to random spots would not have quite the right uniform probability distribution).

When there are no decommissioned nodes (as will be the case pre-HF12) then `total_nodes` and `active_snode_list.size()` will be equal, in which case the double shuffle doesn't happen (and so the results are the same under HF11).